src/Data/Array/Accelerate/Smart.hs

```diff
@@ -185,9 +185,9 @@ import GHC.TypeLits
 --
 -- > mvm_ndp :: Num a => Acc (Matrix a) -> Acc (Vector a) -> Acc (Vector a)
 -- > mvm_ndp mat vec =
--- >   let Z :. rows :. cols  = unlift (shape mat)  :: Z :. Exp Int :. Exp Int
--- >   in  generate (index1 rows)
--- >                (\row -> the $ dotp vec (slice mat (lift (row :. All))))
+-- >   let I2 rows cols  = shape mat
+-- >   in  generate (I1 rows)
+-- >                (\row -> the $ dotp vec (slice mat (row ::. All_)))
```

Comment:
I'm not sure, should this perhaps be written to use an explicit `I1` for the `row`? This would do two things:
1. Avoid `::.`, which you don't really need to use at all until you start defining rank-polymorphic functions
2. Be more transparent and consistent in the treatment of indices. `row` sounds like it's just an integer from 0 to `rows`, but it isn't: it's a `Z :. Int`, and the dimensions just happen to work out so that we can just append `All_`. If the array had been transposed, you would've _needed_ the `I1`.

The code would then look like this:

```haskell
let I2 rows cols = shape mat
in  generate (I1 rows)
             (\(I1 row) -> the $ dotp vec (slice mat (I2 row All_)))
```

Note: I also removed the extraneous space before the `=` in the let-binding; not sure why that was there.